### DISCUSSION TITLE
Milter support

### DISCUSSION THREAD

**Discussion by jonmoesli** [stalwartlabs/smtp-server#1] [Ideas]
Hi, why not implement Milter protocol for DKIM/SPF/ARC validation/signing?
There are great projects, like rspamd, that work very well.

The same idea for SASL authentication, Postfix uses Cyrus/Dovecot for example.

**Comment by mdecimus** [comment]
Hi,

> Hi, why not implement Milter protocol for DKIM/SPF/ARC validation/signing?

Milter is not planned as the goal is that all filtering is done internally by the SMTP. For example, Stalwart SMTP already includes DKIM, SPF, ARC and DMARC support. 
However, support for custom filters will be added through WASM.

> The same idea for SASL authentication, Postfix uses Cyrus/Dovecot for example.

That is already possible. You can SASL authenticate SMTP connections using an IMAP, LMTP or SQL server.

**Reply by bobobo1618** [reply]
README.md [says](https://github.com/stalwartlabs/smtp-server/blob/372d4955d85c97a9723eab1053e213c74b98fdcb/README.md?plain=1#L29):

> Integration with external content filtering systems such as SpamAssassin and ClamAV.

Usually these are integrated using Milter. If they're not implemented using Milter, how can they be integrated?

**Reply by mdecimus** [reply]
Please see here: https://stalw.art/docs/smtp/inbound/data#content-filters

**Reply by jonmoesli** [reply]
What are the limitations of milter? It's a standard implemented in open source and commercial products.

**Reply by mdecimus** [reply]
There are no limitations in the milter protocol. It is currently being developed and should be available next week.

**Reply by jonmoesli** [reply]
Very nice addition! Thank you!

**Comment by mdecimus** [comment]
@jonmoesli @bobobo1618 

Milter support is now available on Stalwart Mail Server v0.3.1!

The official announcement is [here](https://stalw.art/blog/milter-support-smtp) and the updated documentation is [here](https://stalw.art/docs/smtp/filter/milter).

It was tested mainly with `rspamd` so if you have other milter servers please let me know if everything goes well. Thanks.
